### DISCUSSION TITLE
Git storage does not support file rename

### DISCUSSION THREAD

**Discussion by La0** [requarks/wiki#3707] [Error / Bug Report]
**Describe the bug**
When renaming a file in a remote git repository (used as main storage source), the old file remains available in the wiki frontend, and the new file is not available.

**To Reproduce**
1. `git mv a.md b.md`
2. `git commit -a -m 'move a to b'`
3. `git push`
4. wait for wiki to refresh content, nothing happens (look at logs :smile: )

**Expected behavior**
`/a` should not be available, whereas `/b` should be available.

**Host Info (please complete the following information):**
 - OS: Docker (on Debian stable)
 - Wiki.js version: 2.0.12
 - Database engine: postgres 12.1

**Additional context**
I got that warning in the docker logs, the last line looks weird, it's like if the wiki worker tried to process the filename `/wiki/data/repo/a.md => b.md` instead of `/wiki/data/repo/b.md`

```
2020-01-06T21:49:11.876Z [MASTER] info: (STORAGE/GIT) Performing pull rebase from origin on branch master...
2020-01-06T21:49:13.430Z [MASTER] info: (STORAGE/GIT) Performing push to origin on branch master...
2020-01-06T21:49:14.825Z [MASTER] warn: (STORAGE/GIT) Failed to process a.md => b.md
2020-01-06T21:49:14.825Z [MASTER] warn: ENOENT: no such file or directory, open '/wiki/data/repo/a.md => b.md'
```

**Comment by akaRem** [comment]
https://docs.requarks.io/guide/pages#single-character-pages

All single character paths are reserved to access various parts of the Wiki:

**Comment by bmann** [comment]
@akaRem pretty sure that's just example file names.

I think I'm seeing the same thing.

I created a new "2002" folder, and moved all files that were 2002-* into it, and am getting:

```
2020-05-20T08:14:46.882083+00:00 app[web.1]: 2020-05-20T08:14:46.882Z [MASTER] [33mwarn[39m: (STORAGE/GIT) Failed to process archive/{ => 2002}/2002-12-23-boxes-and-arrows-beauty-is-only-screen-deep-sarah-horton.md
2020-05-20T08:14:46.882216+00:00 app[web.1]: 2020-05-20T08:14:46.882Z [MASTER] [33mwarn[39m: ENOENT: no such file or directory, open '/wiki/data/repo/archive/{ => 2002}/2002-12-23-boxes-and-arrows-beauty-is-only-screen-deep-sarah-horton.md'
```

There is no "2002" folder in the WikiJS front end, and the files are still accessible at their original paths.

Which cache can I purge so this is re-read and displays correctly? What is the correct way to move / rename files via git so it is reflected in WikiJS?

BTW: the reason I'm making this move is because I have too many files in one folder, so the front end browse navigation really slows down.

**Comment by spyesx** [comment]
OS: Docker
Wiki.js version: 2.4.107
Database engine: 10.4.8-MariaDB

I've beed working on reorganising our wiki especially from a local clone of the repository. Then I pushed all changes and I can see the same issue here.

- renamed files are not available
- moved files are not available at their new path but still at their old path
- removed items are still available

I've cleared all caches possible from the UI. `/data/repo` is perfectly sync as I can see the same structure. (That's reassuring 😄 )  "Rerender All Pages" and "Rebuild Page Tree" from Utilities don't change anything.

I thought it could come from inner links so we removed them all to switch to a "Browse" navigation only. No change.

Is there a way to rebuild the entire wiki tree from the repo?

I'd be happy to provide more informations to help if you have specific questions.

**Comment by tionis** [comment]
This is the same bug as #1440.

**Comment by fpagnoux** [comment]
I'm also hitting this limitation, and that's really unfortunate. 

What I really liked about Wiki.js is the ability to keep the wiki in sync with git, so that changes could be made either from the web UI or directly in the repo. But that doesn't work well in 2.x: it's not clear when will changes in git will be reflected in the UI (if they are at all), and I find myself very regularly trying to rebuild the tree or re-render all the pages.

**Comment by jantari** [comment]
We also ran into this problem when evaluating wiki.js - unfortunately I think this will make the entire git integration and therefore wiki rather pointless to us

A simple way to reproduce from a clean state is:

1. Set up bidirectional sync with git
2. In the git repo:
   ```
   mkdir testmove
   echo '<a href="page2.html">Click here</a>' > testmove/page1.html
   echo '<h1>Hello!</h1>' > testmove/page2.html
    ```
3. git add, commit and push - wait for wikijs to sync and the pages to show up, verify the "Click here" link works

4. In the git repo:
    `mv page2.html page_newname.html`
    Then update the link in page1.html to point to the new filename, using an editor or sed etc.:
    `sed -i "s/page2/page_newname/g" page1.html`
6. git diff should look like this (after git add .):
    ```
    $ git diff --staged
    testmove/page1.html
    
     1  ⋮    │<a href="page2.html">Click here</a>
        ⋮ 1  │<a href="page_newname.html">Click here</a>
    
    renamed: testmove/page2.html -> testmove/page_newname.html
    ```
5. git add, commit and push - wait for wikijs to sync
6. Navigate to page1, click the link
7. It doesn't work anymore because the rename of page2.html to page_newname.html is not reflected in wikijs!

**Comment by EricFromCanada** [comment]
I've [submitted a pull request](https://github.com/Requarks/wiki/pull/4307) to address this; feel free to try it out. For now, file renames need to occur in Git as separate commits to remove and add the file in the old and new locations.

**Comment by peteruithoven** [comment]
I'm running Wiki.js version 2.5.268 but renames in git don't seem to get synced.

1. I create a new page on the wiki (rename-test)
2. Force sync
3. File appears in git repo. (rename-test.md)
4. I edit the contents of the page's file in the git repo using commit, push etc
5. Force sync
5. Page contents is updated (So I'm sure I can edit things from git)
6. I rename the file to rename-test-renamed.md in the git repo using commit, push etc
7. Force sync
8. Old page is still there, the path is still /rename-test. There is no /rename-test-renamed.

Also, creating a page from git works just fine.

**Reply by peteruithoven** [reply]
After renaming a file locally I also tried committing a delete and then committing adding the renamed file.  
Even after a force sync, "Rebuild Page Tree" and "Flush Pages and Assets Cache" this has 0 effect on the online wiki.

**Reply by peteruithoven** [reply]
Found one possible workaround, when I do a force sync after the delete and then another after the add the change does seem to be picked up. Make sure that after syncing the delete step the wiki is actually updated. 
Otherwise you'll need to use "Add Untracked Changes" to bring the wiki's git back in sync with the DB again. 

All steps:
1. Move folder / file you want to rename out of your local git repo
2. Commit the deletion
3. Push
4. Force sync in Wiki admin
5. Check wiki that folder / file is actually gone
6. Move the folder / file back into the local git repo
7. Rename the folder / file
8. Commit the "new" files
9. Push
10. Force sync in Wiki admin
11. File / folder should be back & renamed.
